src/plugins/rgbcolor/rgbcolor.c

```diff
@@ -0,0 +1,379 @@
+		{
+			expandedStr[2 * i - 1] = str[i];
+			expandedStr[2 * i] = str[i];
+		}
+	}
+	else
+	{
+		strcpy (expandedStr, str);
+	}
+
+	if (colVar == HEX_THREE || colVar == HEX_SIX)
+	{
+		expandedStr[7] = 'f';
+		expandedStr[8] = 'f';
+	}
+
+	expandedStr[9] = '\0';
+
+	return expandedStr;
+}
+
+static void elektraColorNormalizeHexString (Key * key, ColorVariant colVar)
+{
+	const char * str = keyString (key);
+	char * origvalue = elektraStrDup (str);
+
+	kdb_unsigned_long_t color;
+	if (colVar == NAMED_COLOR)
+	{
+		const NamedColor * namedColor = findNamedColor (str);
+		color = namedColor->value;
+	}
+	else if (colVar != HEX_EIGHT)
+	{
+		char * expandedStr = elektraColorExpand (str, colVar);
+		ELEKTRA_LOG_DEBUG ("Expanded %s to %s", str, expandedStr);
+		color = ELEKTRA_UNSIGNED_LONG_LONG_S (expandedStr + 1, NULL, 16);
+		elektraFree (expandedStr);
+	}
+	else
+	{
+		color = ELEKTRA_UNSIGNED_LONG_LONG_S (str + 1, NULL, 16);
+	}
+	elektraColorSetInteger (key, color);
+
+	ELEKTRA_LOG_DEBUG ("Setting origvalue of %s to %s", keyName (key), origvalue);
+	keySetMeta (key, "origvalue", origvalue);
+	elektraFree (origvalue);
+}
+
+static void elektraColorRestore (Key * key)
+{
+	const Key * orig = keyGetMeta (key, "origvalue");
+	if (orig != NULL)
+	{
+		ELEKTRA_LOG_DEBUG ("Restore %s to %s", keyName (key), keyString (orig));
+		keySetString (key, keyString (orig));
+	}
+}
+
+int elektraRgbcolorGet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned, Key * parentKey)
+{
+	if (!elektraStrCmp (keyName (parentKey), "system/elektra/modules/rgbcolor"))
+	{
+		KeySet * contract =
+			ksNew (30, keyNew ("system/elektra/modules/rgbcolor", KEY_VALUE, "rgbcolor plugin waits for your orders", KEY_END),
+			       keyNew ("system/elektra/modules/rgbcolor/exports", KEY_END),
+			       keyNew ("system/elektra/modules/rgbcolor/exports/get", KEY_FUNC, elektraRgbcolorGet, KEY_END),
+			       keyNew ("system/elektra/modules/rgbcolor/exports/set", KEY_FUNC, elektraRgbcolorSet, KEY_END),
+#include ELEKTRA_README
+			       keyNew ("system/elektra/modules/rgbcolor/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END), KS_END);
+		ksAppend (returned, contract);
+		ksDel (contract);
+
+		return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	}
+
+	Key * cur;
+	ksRewind (returned);
+	while ((cur = ksNext (returned)) != NULL)
+	{
+		const Key * meta = keyGetMeta (cur, "check/rgbcolor");
+		if (!meta) continue;
+		ColorVariant colVar = is_valid_key (cur, parentKey);
+		elektraColorNormalizeHexString (cur, colVar);
+	}
+
+	return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+}
+
+int elektraRgbcolorSet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned ELEKTRA_UNUSED, Key * parentKey ELEKTRA_UNUSED)
+{
+	// set all keys
+	// this function is optional
+	Key * cur;
+	ksRewind (returned);
+	while ((cur = ksNext (returned)) != NULL)
+	{
+		const Key * meta = keyGetMeta (cur, "check/rgbcolor");
+		if (!meta) continue;
+		ColorVariant colVar = is_valid_key (cur, parentKey);
```

Comment:
I think you don't need this first call, if you call `elektraColorRestore` unconditionally.

Reply:
Great catch, I removed the first call.